backend/tests/test_b07_p3_provider_controls.py

```
@@ -129,6 +129,121 @@ async def test_p3_hourly_shutoff_distinct_from_monthly(monkeypatch, test_tenant)
         assert any(row.is_shutoff for row in shutoff_rows)
 
 
+@pytest.mark.asyncio
+async def test_p3_kill_switch_blocks_without_provider_attempt_non_vacuous(monkeypatch, test_tenant):
+    monkeypatch.setattr(settings, "LLM_HOURLY_SHUTOFF_CENTS", 10_000, raising=False)
+    monkeypatch.setattr(settings, "LLM_MONTHLY_CAP_CENTS", 10_000, raising=False)
+
+    calls = {"count": 0}
+
+    async def _spy_provider(*, requested_model, prompt, reservation):
```

Ruff (0.15.0)

[warning] Missing return type annotation for private function `_spy_provider`

(ANN202)